### PR TITLE
PERF: improve conversion to BooleanArray from int/float array

### LINKED ISSUE #29838: PERF: improve conversion to BooleanArray from int/float array
Currently, the creation of a BooleanArray from an int/float array goes through a conversion to object dtype (to do it together with the generic conversion from any list-like):

https://github.com/pandas-dev/pandas/blob/7d7f885856b0c3d51eaf15beaac9d4f30c23797d/pandas/core/arrays/boolean.py#L133-L145

For the specific case of int/float ndarray, this could be optimized with a specific path for those cases without the casting to object array (probably just skipping the `np.asarray(values, dtype=object)` if `values` is a float/int ndarray will be enough).

### PR DESCRIPTION
**Pull request by ethanywang**
- [x] closes pandas-dev#29838
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/arrays/boolean.py [hunk 1]
```diff
@@ -130,9 +130,18 @@ def coerce_to_array(values, mask=None, copy: bool = False):
     if isinstance(values, np.ndarray) and values.dtype == np.bool_:
         if copy:
             values = values.copy()
+    elif isinstance(values, np.ndarray) and values.dtype in (np.int_, np.float_):
+        values_copy = values.copy()
```
Comment (jreback) [inline pandas/core/arrays/boolean.py]: i don't think you actually need to copy here

**Review by jreback — CHANGES_REQUESTED** [review]
can you add this issue number onto the list for adding boolean array in the whatsnew.

can you add some tests specifically for this (we *might*  have them, but pls point them out)

**Comment by ethanywang** [comment]
Related Test:
https://github.com/pandas-dev/pandas/blob/2d5455cbb14c39093396c6a21d67c99dc52175e9/pandas/tests/arrays/test_boolean.py#L136-L155

https://github.com/pandas-dev/pandas/blob/2d5455cbb14c39093396c6a21d67c99dc52175e9/pandas/tests/arrays/test_boolean.py#L108

File: pandas/core/arrays/boolean.py [hunk 2]
```diff
@@ -130,9 +130,19 @@ def coerce_to_array(values, mask=None, copy: bool = False):
     if isinstance(values, np.ndarray) and values.dtype == np.bool_:
         if copy:
             values = values.copy()
+    elif isinstance(values, np.ndarray) and values.dtype in (np.int_, np.float_):
```
Comment (groutr) [inline pandas/core/arrays/boolean.py]: I think this check is incorrect.
```python
a = np.array([1], dtype=np.int8)
a.dtype == np.int_   # False
```

Should this be using the dtype checking functions from pandas.dtypes.common (like `is_numeric_dtype`)?

Comment (WillAyd) [inline pandas/core/arrays/boolean.py]: I *think* should just use `is_numeric_dtype` here instead

**Comment by jreback** [comment]
ok lgtm. can you do a simple benchmark on this and show it (+1 if you can add it to the asvs)

**Review by WillAyd — APPROVED** [review]
lgtm as well outside of @jreback comments

**Comment by ethanywang** [comment]
I used pytest-benchmark to do the simple benchmarking, as I'm not quite sure where should I write the benchmark for the boolean array in the asv folder.

Original Branch:
```
------------------------------------------------------------------------------------------------ benchmark: 2 tests ------------------------------------------------------------------------------------------------
Name (time in us)                          Min                   Max                Mean              StdDev              Median                 IQR            Outliers  OPS (Kops/s)            Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_benchmark_from_integer_array     632.0190 (1.0)      4,040.8400 (1.0)      959.2070 (1.11)     343.6619 (1.04)     859.9870 (1.13)     310.8663 (1.28)        86;37        1.0425 (0.90)        829           1
test_benchmark_from_float_array       650.8890 (1.03)     4,129.7300 (1.02)     863.9706 (1.0)      330.2377 (1.0)      761.4645 (1.0)      242.4200 (1.0)         96;64        1.1574 (1.0)         978           1
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

New Branch:
```
------------------------------------------------------------------------------------------------ benchmark: 2 tests -----------------------------------------------------------------------------------------------
Name (time in us)                          Min                   Max                Mean              StdDev              Median                IQR            Outliers  OPS (Kops/s)            Rounds  Iterations
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_benchmark_from_float_array       580.4190 (1.0)      2,836.9650 (1.53)     687.1475 (1.02)     196.7836 (1.69)     611.1900 (1.0)      97.8967 (2.22)      117;146        1.4553 (0.98)       1359           1
test_benchmark_from_integer_array     609.3710 (1.05)     1,856.8680 (1.0)      675.1754 (1.0)      116.3234 (1.0)      637.5215 (1.04)     44.0175 (1.0)         62;95        1.4811 (1.0)         760           1
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

Is it now okay to merge? @jreback

**Comment by jreback** [comment]
@ethanywang see the asv docs here: https://dev.pandas.io/docs/development/contributing.html#running-the-performance-test-suite

can you construct some asvs which add some benchmarks (and then show the results here)

**Comment by ethanywang** [comment]
@jreback Using the asv bechmark. The results are:
```
       before           after         ratio
     [c0f6428b]       [3dcbbd61]
     <master>         <int_float_to_boolean>
-        90.7±1μs       60.2±0.7μs     0.66  array.BooleanArray.time_from_float_array
-        95.0±4μs         60.5±2μs     0.64  array.BooleanArray.time_from_integer_array

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
perfect @ethanywang 

would a follow up PR for similar asvs for IntegerArray and StringArray

(we might have some in series somewhere for IntegerArray already)
can move those

**Comment by ethanywang** [comment]
@jreback So you mean I can remove the array.py in the asv_benchmark folder, and not submit it in this PR?

**Comment by jreback** [comment]
> @jreback So you mean I can remove the array.py in the asv_benchmark folder, and not submit it in this PR?

no in a follow up PR i would like to add asv constructions for IntegerArray and StringArray in array.py

we may have some construction benchmarks already for Integer dtypes in Series which we can move

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
@ethanywang Thanks a lot!

**Pull request merged**